### PR TITLE
Advanced Druid Client and Client API

### PR DESCRIPTION
**Pull request by anskarl**
- Adds API for implementing custom clients
  - Adds Advanced Druid client with cached host connection pool, load balancing and queue
  - Updates configuration to support multiple Druid Brokers and custom configuration for Clients
  - Adds health check per Broker
  - Adds and updates tests
  - Adds docker-compose with ClusterF images for unreliable HTTP proxy in order to simulate failures
  - Updates CircleCI config to use docker-compose
  - Adds tests for configuration overrides
  - Adds documentation for Druid Clients and their API
  - Adjusts timeouts in unit tests
  - Adds a configuration builder for DruidAdvancedHttpClient, in order to programmatically define the parameters of the client

GH-61

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  scruid_druid:
+    image: mmartina/docker-druid:latest
+    container_name: scruid_druid
+    ports:
+      - "8081:8081"
+      - "8082:8082"
+      - "8083:8082"
+      - "8084:8082"
+
+  wait_for_druid:
```
Comment (Fokko) [inline docker-compose.yml]: Maybe we can do this with a startup script as well. I'm hesitant to introduce another container for only waiting for the services to come up. Something similar to https://github.com/puckel/docker-airflow/blob/master/script/entrypoint.sh#L45-L57

WDYT?

Comment (anskarl) [reply docker-compose.yml]: sure, I have updated `docker-compose.yml` and added `services.sh` script file that follows the approach that you suggest ([https://github.com/puckel/docker-airflow/blob/master/script/entrypoint.sh#L45-L57](https://github.com/puckel/docker-airflow/blob/master/script/entrypoint.sh#L45-L57)).

File: README.md [hunk 2]
```diff
@@ -137,15 +135,17 @@ The configuration is done by [Typesafe config](https://github.com/typesafehub/co
...
+  hosts = ${?DRUID_HOSTS}
   secure = false
   secure = ${?DRUID_USE_SECURE_CONNECTION}
   url = "/druid/v2/"
   url = ${?DRUID_URL}
-
+  health-endpoint = "/status/health"
+  health-endpoint = ${?DRUID_HEALTH_ENDPOINT}
+  client-backend = "ing.wbaa.druid.client.DruidHttpClient"
+  client-backend = ${?DRUID_CLIENT_BACKEND
```
Comment (bjgbeelen) [inline README.md]: There is a `}` missing here

Comment (anskarl) [reply README.md]: fixed :)

Comment (bjgbeelen) [reply README.md]: I wasnn't aware I already sent this comment :) will continue the review now

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    image: mmartina/docker-druid:latest
+    container_name: scruid_druid
+    ports:
+      - "8081:8081"
+      - "8082:8082"
+      - "8083:8082"
+      - "8084:8082"
+
+  delay_response:
+    image: andymacdonald/clusterf-chaos-proxy:latest
```
Comment (bjgbeelen) [inline docker-compose.yml]: this is pretty cool!

File: src/test/scala/ing/wbaa/druid/DruidAdvancedHttpClientSpec.scala [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        // the following node always fails with Internal Server Error (HTTP code 500)
+        QueryHost("localhost", 8086) -> false
+      )
+
+      val config = DruidConfig(clientBackend = classOf[DruidAdvancedHttpClient],
+                               hosts = expectedHealthCheck.keys.toSeq)
+
+      val client = config.client
+
+      // since localhost:8087 is always failing the health status should be false
```
Comment (bjgbeelen) [inline src/test/scala/ing/wbaa/druid/DruidAdvancedHttpClientSpec.scala]: very minor, this should be 8086?

File: services.sh [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    if [ $j -ge $TRY_LOOP ]; then
+      echo >&2 "INFO $(date) - $host:$port still not reachable, giving up"
+      exit 1
+    fi
+    echo "INFO $(date) - waiting for $name... $j/$TRY_LOOP"
+    sleep 5
+  done
+}
+
+foo(){
```
[the remark this answers is not in the document]
Comment (anskarl) [reply services.sh]: oups! my mistake, I should have delete this function.

File: src/test/scala/ing/wbaa/druid/DruidAdvancedHttpClientSpec.scala (hunk 4 above)
Comment (anskarl) [reply src/test/scala/ing/wbaa/druid/DruidAdvancedHttpClientSpec.scala]: I will fix it

**Pull request merged**